TEP-0118: Enable pipeline to handle matrix include params

EmmaMunley opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

[[TEP-0090: Matrix](https://github.com/tektoncd/community/blob/main/teps/0090-matrix.md)] introduced `Matrix` to the `PipelineTask` specification such that the `PipelineTask` executes a list of `TaskRuns` or `Runs` in parallel with the specified list of inputs for a `Parameter` or with different combinations of the inputs for a set of `Parameters`.

To build on this [TEP-0018 ](https://github.com/tektoncd/community/blob/main/teps/0118-matrix-with-explicit-combinations-of-parameters.md )introduced Matrix.Include, which allows passing in a specific combinations of `Parameters` into the `Matrix`.

**This PR enables pipeline to handle matrix include params.**

Note: This feature is still in preview mode. Implementation logic will be added in subsequent commits.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

JeromeJu commented:
/assign

QuanZhang-William commented:
/assign

QuanZhang-William commented on pkg/apis/pipeline/v1/pipeline_validation_test.go:
```diff
@@ -3252,7 +3252,170 @@ func Test_validateResultsFromMatrixedPipelineTasksNotConsumed(t *testing.T) {
 			Paths:   []string{"tasks[1]", "finally[0]"},
 		},
-	}}
+	}, {
```
Did we add validation logic in the PR that needs to be covered by these tests?

QuanZhang-William reviewed (commented):
I think in general we need to make sure the logic and tests are synced for both v1 and v1beta1 API versions at this time?

EmmaMunley commented:
/hold Thanks Quan! Holding this for now. May need to include implementation FanOut() logic with this as well.

JeromeJu commented on pkg/apis/pipeline/v1/matrix_types.go:
```diff
@@ -110,7 +110,7 @@ func (m *Matrix) CountCombinations() int {
 // parameters
 func (m *Matrix) countGeneratedCombinationsFromParams() int {
-	if !m.hasParams() {
+	if !m.HasParams() {
```
Thanks for the changes @EmmaMunley , can we elaborate the reasons why we made those functions exported? and if so might as well update it in the PR message? 🙏

JeromeJu commented on pkg/apis/pipeline/v1beta1/pipeline_conversion.go:
```diff
@@ -259,13 +259,39 @@ func (m *Matrix) convertTo(ctx context.Context, sink *v1.Matrix) {
+		}
+	}
+
+	if source.HasInclude() {
```
we might not need the `HasInclude` condition check here

EmmaMunley replied:
Thanks! Removed

JeromeJu commented:
I am not fully clear with the title, can I confirm it is adding the conversion for `matrix.include` and refactoring the validation functions?
thanks 🙏

JeromeJu commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -176,6 +176,11 @@ func ApplyTaskResults(targets PipelineRunState, resolvedResultRefs ResolvedResul
 			pipelineTask.Params = replaceParamValues(pipelineTask.Params, stringReplacements, arrayReplacements, objectReplacements)
 			if pipelineTask.IsMatrixed() {
 				pipelineTask.Matrix.Params = replaceParamValues(pipelineTask.Matrix.Params, stringReplacements, nil, nil)
+				if pipelineTask.Matrix.HasInclude() {
```
this felt to be changes for the reconciler, which doesn't feel quite related to the current PR title/description 🤔

EmmaMunley commented on pkg/apis/pipeline/v1beta1/pipeline_conversion_test.go:
```diff
@@ -157,12 +164,10 @@ func TestPipelineConversion(t *testing.T) {
 					t.Errorf("ConvertTo() = %v", err)
 					return
 				}
-				t.Logf("ConvertTo() = %#v", ver)
```
That was by mistake :)

JeromeJu commented on pkg/apis/pipeline/v1beta1/pipeline_conversion.go:
```diff
@@ -259,13 +259,37 @@ func (m *Matrix) convertTo(ctx context.Context, sink *v1.Matrix) {
 		m.Params = append(m.Params, new)
+		if m.hasInclude() {
+			for _, include := range m.Include {
+				include.Params = append(include.Params, new)
```
This feels like not changing the actuall Matrix object. so we could remove the iteration here

EmmaMunley commented:
/ unhold

EmmaMunley commented:
/ unhold

JeromeJu commented:
/unhold
(the command needs to be right after the slash)

JeromeJu commented on pkg/apis/pipeline/v1beta1/pipeline_conversion.go:
```diff
@@ -259,6 +259,16 @@ func (m *Matrix) convertTo(ctx context.Context, sink *v1.Matrix) {
 		param.convertTo(ctx, &new)
 		sink.Params = append(sink.Params, new)
 	}
+	if m.hasInclude() {
```
nit: the check here might also be removed

jerop reviewed (changes requested):
"handling matrix include params" is ambiguous, let's clarify what this means in the commit and PR description 🙏🏾 

looking into the PR itself, it seems there are three things happening:
- refactoring extraction of parameters from a pipelinetask
- conversion of matrix.include to/from v1
- substituting result references in matrix.include.params

I recommend splitting this PR into three PRs such that each does only one thing 🙏🏾

lbernick commented:
@EmmaMunley related to Jerop's comment, with a lot of background copy-pasted between matrix include commit messages, it's hard to tell what's actually changing as a result of the commit. Can you focus the commit messages more on the specific changes being introduced?

EmmaMunley commented:
Thank you for the feedback and suggestions. @JeromeJu @JeromeJu  @lbernick 

I have addressed the comments and broken this out into multiple PRs and shortened the commit messages to be more specific. 

- Refactor extracting parameters from a PipelineTask (#6344 )
- TEP-0118: Apply Replacements for Matrix Include Parameters (#6345)
- TEP-0118: Update Pipeline Conversion for Matrix Include Parameters(https://github.com/tektoncd/pipeline/pull/6346)
- TEP-0118: Update PipelineTaskResultRefs for Matrix Include Parameters(#6348)

cc: @QuanZhang-William

JeromeJu commented:
thanks for breaking the PRs up Emma!
